Adds decorate functionality to module output (#1357)
insightsengineering/teal: vignettes/decorate-modules-output.Rmd

`server` components, where they will be used by `ui/srv_teal_transform_module`:

```{r}
tm_decorated_plot <- function(label = "module", transforms = list(), decorators = teal_transform_module()) {
  module(
    label = label,
    ui = function(id, decorators) {
      ns <- NS(id)
      div(
        selectInput(ns("dataname"), label = "select dataname", choices = NULL),
        selectInput(ns("x"), label = "select x", choices = NULL),
        selectInput(ns("y"), label = "select y", choices = NULL),
        ui_teal_transform_module(ns("decorate"), transforms = decorators),
        plotOutput(ns("plot")),
        verbatimTextOutput(ns("text"))
      )
    },
    server = function(id, data, decorators) {
      moduleServer(id, function(input, output, session) {
        observeEvent(data(), {
          updateSelectInput(inputId = "dataname", choices = teal.data::datanames(data()))
        })

        observeEvent(input$dataname, {
          updateSelectInput(inputId = "x", choices = colnames(data()[[input$dataname]]))
          updateSelectInput(inputId = "y", label = "select y", choices = colnames(data()[[input$dataname]]))
        })

>         q1 <- reactive({

Reviewer: Invent a better name for `q1`?

                  ggplot2::geom_point()
              },
              dataname = as.name(input$dataname),
              x = as.name(input$x),
              y = as.name(input$y)
            )
        })

        q2 <- srv_teal_transform_module("decorate", data = q1, transforms = decorators)

        plot_r <- reactive({
          req(q2())
          q2()[["plot"]]
        })

        output$plot <- renderPlot(plot_r())
        output$text <- renderText({
          teal.code::get_code(q2())
        })
      })
    },
    ui_args = list(decorators = decorators),
    server_args = list(decorators = decorators)
  )
}
```

## Example app

```{r}
library(ggplot2)
app <- init(
  data = teal_data(iris = iris, mtcars = mtcars),
  modules = modules(
    tm_decorated_plot("identity"),
    tm_decorated_plot("no-ui", decorator = static_decorator),
    tm_decorated_plot("lang", decorator = static_decorator_lang),
    tm_decorated_plot("interactive", decorator = interactive_decorator),
    tm_decorated_plot("interactive-from lang", decorator = interactive_decorator_lang),
    tm_decorated_plot("from-fun", decorator = gg_xlab_decorator("plot")),

>     tm_decorated_plot("failing", decorator = failing_decorator)

Reviewer: Those decorators are not created in this vignette. Are they distributed with `teal`? Should we include them here

  )
)

if (interactive()) {
  shinyApp(app$ui, app$server)
}
```

## To improve

- Error handling when a decorator throws an error.
- Rewrite modules so that they always have a graph as `graph` (or `plot`, `g`, `p` etc.) and table as 
  `table` (or `t` etc.) and listings as `listings` (or `l` etc.). This simplifies/standardizes a construction of

>   `teal_transform_module` which can be re-used in multiple modules. 

Reviewer: This should be removed.

The rewrite of objects in `teal.modules.*` should be handled as an extra task. Include: 

- `teal.modules.general`
- `teal.modules.clinical`
- `teal.osprey`
- `teal.helios`

---

> title: "Customizing module output"

Reviewer: Customizing or Decorating
```suggestion
title: "Decorating module output"
```

Reviewer: Change file name if we decide on the final name

      moduleServer(id, function(input, output, session) {
        reactive({
          req(data())
          data() |> within(
            {
              output_name <- output_name +
                xlab(x_axis_title)
            },
            x_axis_title = input$x_axis_title,
            output_name = as.name(output_name)
          )
        })
      })
    }
  )
}
    ```

Decorator failures are managed by an internal `teal` mechanism called `trigger_on_success`, which ensures that the `data`
object within the module remains intact. If a decorator fails, it will be ignored, and an appropriate error message will be displayed.

```{r}
failing_decorator <- teal_transform_module(
  ui = function(id) {
    ns <- NS(id)
    div(
      textInput(ns("x_axis_title"), "X axis title", value = "x axis")
    )
  },
  server = function(id, data) {
    moduleServer(id, function(input, output, session) {
      reactive(stop("This is error"))
    })
  }
)
    ```

## Example Module

To include decorators in a `teal` module, pass them as arguments (`ui_args` and `server_args`) to the module’s `ui` and 
`server` components, where they will be used by `ui/srv_teal_transform_module`:

```{r}

> tm_decorated_plot <- function(label = "module", transforms = list(), decorators = teal_transform_module()) {

Reviewer: It would be great if `transforms` and `decorators` had the same default/empty input type.
Is it necessary that `decorators` need to be an empty `teal_transform_module()`? 
If it could be `list()` that would be consistent with `transforms`

---
title: "Customizing Module Output"
author: "NEST CoreDev"
output:
  rmarkdown::html_vignette:
    toc: true
vignette: >
  %\VignetteIndexEntry{Customizing Module Output}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

## Introduction

`teal` is a powerful `shiny`-based framework with built-in modules for interactive data analysis. 
This document outlines the customization options available for modifying the output of `teal` modules. 

> You will learn how to use `teal_transform_module` to modify and enhance the objects created by `teal` modules, 

Author: "modules" is a shiny term and it refers to any module. Decorate is dedicated to modify objects in `teal_module` - these created by `module()` function.

Reviewer: Should I rephrase to 

> `teal` is a powerful `shiny`-based framework with built-id tools for interactive data analysis. 
This document outlines the customization options available for modifying the output of `teal` modules. 
You will learn how to use `teal_transform_module` to modify and enhance the objects created by `teal::module()`,

Reviewer: 60eec6e8fc66db73d4d4a56d4979d2dfee59580f

enabling you to tailor the outputs to your specific requirements without rewriting the original module code.

Adjusting input data or customizing module outputs in `teal` is accomplished using `transforms` created through 
`teal_transform_module`. This document will focus on modifying outputs, and for clarity, we will refer to these 
output-modifying objects as decoration objects or decorators.

> Throughout the examples, you will see references to a `data()` object. 

Author: `data()` has been extensively described in other vignette Including Data in teal Applications. I don't know if it is worth it to explain it again here

Reviewer: Cool, will remove

Reviewer: f9bdb60c692ffc465975f747232eee164b36a2d9

This reactive object, built with `teal_data()`[https://insightsengineering.github.io/teal.data/latest-tag/articles/teal-data.html],

> holds all the data used `within` the module as well as the generated output objects. 
> `teal_transform_module` operates on this object and can execute custom R code within the `data()` environment using the `within function`. 

Author: Too many within(s) used in different meaning.
 
```suggestion
holds all the data used inside the module as well as the generated output objects. 
`teal_transform_module` operates on this object and can execute custom R code in the `data()` environment using the `within()`. 
```

Reviewer: f9bdb60c692ffc465975f747232eee164b36a2d9

This approach allows you to pass code directly and evaluate it in the `data()` environment.

## Decorators

### Server

> The simplest way to create a decorator is to use `teal_transform_module` without specifying a `ui`.

Author: should be have a `teal_decorator_module`?

Reviewer: Unsure, I think using `teal_transform_module` to show that it reuses some logic is OK

Reviewer: And if we have `teal_transform_module` (with a verb in the name) then for decorators we should have `teal_decorate_module`

Reviewer: But maybe with `teal_decorate_module` it would be easier to document and explain and show the different purpose, even though it's quite similar. I'm puzzled :D

        ggtitle("This is title") +
        xlab(x_axis_title)
    )
  )
)
```

## Handling Various Object Names

`teal_transform_module` relies on the names of objects created within a module. 
Writing a decorator that applies to any module can be challenging because different modules may use different object names.
It is recommended to create a library of decorator functions that can be adapted to the specific object names used in `teal` modules.
In the following example, focus on the `output_name` parameter to see how decorator can be applied to multiple modules:

```{r}
gg_xlab_decorator <- function(output_name) {
  teal_transform_module(
    ui = function(id) {
      ns <- NS(id)
      div(
        textInput(ns("x_axis_title"), "X axis title", value = "x axis")
      )
    },
    server = function(id, data) {
      moduleServer(id, function(input, output, session) {
        reactive({
          req(data())
          within(data(),
            {
              output_name <- output_name +
                xlab(x_axis_title)
            },
            x_axis_title = input$x_axis_title,
            output_name = as.name(output_name)
          )
        })
      })
    }
  )
}
```

> Decorator failures are managed by an internal `teal` mechanism called `trigger_on_success`, which ensures that the `data`

Author: Original function name is `.trigger_on_success` but I would try to avoid giving the internal name - to follow the rule to never expose any internals in the public documentation
 ```suggestion
Decorator failures are managed by an internal `teal` mechanism called "trigger on success", which ensures that the `data`
```

Reviewer: f9bdb60c692ffc465975f747232eee164b36a2d9

        plot_r <- reactive({
          req(plot_data_decorated())
          plot_data_decorated()[["plot"]]
        })

        output$plot <- renderPlot(plot_r())
        output$text <- renderText({
          teal.code::get_code(req(plot_data_decorated()))
        })
      })
    },
    ui_args = list(decorators = decorators),
    server_args = list(decorators = decorators)
  )
}
```

### Application

```{r}
library(ggplot2)
app <- init(
  data = teal_data(iris = iris, mtcars = mtcars),
  modules = modules(
    tm_decorated_plot("identity"),
    tm_decorated_plot("no-ui", decorators = static_decorator),
    tm_decorated_plot("lang", decorators = static_decorator_lang),
    tm_decorated_plot("interactive", decorators = interactive_decorator),
    tm_decorated_plot("interactive-from lang", decorators = interactive_decorator_lang),
    tm_decorated_plot("from-fun", decorators = gg_xlab_decorator("plot")),
    tm_decorated_plot("failing", decorators = failing_decorator)
  )
)

if (interactive()) {
  shinyApp(app$ui, app$server)
}
```

## Decorating Tables

### Example Module

```{r, eval=requireNamespace("DT")}
library(DT)

# Define a decorator module for customizing table appearance

> custom_table_decorator <- teal_transform_module(

Author: I think `custom_table_decorator` should occupy separate chunk and be described

Reviewer: f9bdb60c692ffc465975f747232eee164b36a2d9
